### PR TITLE
Support for route tables delegation selector expressions

### LINKED ISSUE #3085: delegated weighted routes are not sorted globally
**Describe the bug**
A route of higher weight (10) can override a route of lesser weight (0-9) depending the routeDelegateAction ordering within the virtualservice.

**To Reproduce**
Create a virtual service with 2 delegations for the same prefix matcher, with different routeTableSelectors, example with simple different namespace selectors:

```yaml
---
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: foo
  namespace: gloo-system
spec:
  virtualHost:
    domains:
      - foo.io
    routes:
      - matchers:
          - prefix: /
        delegateAction:
          selector:
            namespaces:
              - "ns1"
      - matchers:
          - prefix: /
        delegateAction:
          selector:
            namespaces:
              - "ns2"
```

Create the following routeTables:
```yaml
---
apiVersion: gateway.solo.io/v1
kind: RouteTable
metadata:
  name: first
  namespace: ns1
spec:
  routes:
    - matchers:
        - prefix: /foo/a/
      routeAction:
        single:
          upstream:
            name: ns1-first-80
            namespace: gloo-system
---
apiVersion: gateway.solo.io/v1
kind: RouteTable
metadata:
  name: second
  namespace: ns2
spec:
  // also happens with default weight of 0
  weight: 5
  routes:
    - matchers:
        - prefix: /foo/b/
      routeAction:
        single:
          upstream:
            name: ns2-second-80
            namespace: gloo-system
---
apiVersion: gateway.solo.io/v1
kind: RouteTable
metadata:
  name: third
  namespace: ns1
spec:
 weight: 10
  routes:
    - matchers:
        - prefix: /
      routeAction:
        single:
          upstream:
            name: ns1-third-80
            namespace: gloo-system
```

Observe  the resulting route ordering in the proxy resource (first, third, second):
```
glooctl get proxy -o yaml 
```

**Expected behavior**

The ordering should be first, second, third

**Additional context**

- Gloo version 1.3.9
- Kubernetes version 1.16.8-gke.15

Unit tests that helped me confirm this behavior is done on purpose even on latest Gloo (master branch / 1.4.0-beta12 tag), add in `translator_test` package, in file `projects/gateway/pkg/translator/converter_test.go`) 

```go
Describe("route tables with weights and different label selectors", func() {

			var rt1a, rt1b, rt2 *v1.RouteTable

			buildVirtualServiceWithMultipleRoutes := func(routes []*v1.Route) *v1.VirtualService {
				return &v1.VirtualService{
					Metadata: core.Metadata{
						Name:      "vs-1",
						Namespace: "ns-1",
					},
					VirtualHost: &v1.VirtualHost{
						Routes: routes,
					},
				}
			}

			BeforeEach(func() {

				// Matches rt1a, rt1b, rt2
				vs = buildVirtualServiceWithMultipleRoutes([]*v1.Route{
					{
						Matchers: []*matchers.Matcher{{
							PathSpecifier: &matchers.Matcher_Prefix{
								Prefix: "/foo/",
							},
						}},
						Action: &v1.Route_DelegateAction{
							DelegateAction: &v1.DelegateAction{
								DelegationType: &v1.DelegateAction_Selector{
									Selector: &v1.RouteTableSelector{
										Labels:     map[string]string{"foo": "bar"},
										Namespaces: []string{"*"},
									},
								},
							},
						},
					},
					{
						Matchers: []*matchers.Matcher{{
							PathSpecifier: &matchers.Matcher_Prefix{
								Prefix: "/foo/",
							},
						}},
						Action: &v1.Route_DelegateAction{
							DelegateAction: &v1.DelegateAction{
								DelegationType: &v1.DelegateAction_Selector{
									Selector: &v1.RouteTableSelector{
										Labels:     map[string]string{"bar": "baz"},
										Namespaces: []string{"*"},
									},
								},
							},
						},
					},
				})

				// No weight
				rt1a = buildRouteTableWithSimpleAction("rt-1a", "ns-1", "/foo/a/", map[string]string{"foo": "bar"})

				// heavy weight
				rt1b = buildRouteTableWithSimpleAction("rt-1b", "ns-1", "/foo/", map[string]string{"foo": "bar"})
				rt1b.Weight = &types.Int32Value{Value: 10}

				// middle weight
				rt2 = buildRouteTableWithSimpleAction("rt-2", "ns-2", "/foo/b/", map[string]string{"bar": "baz"})
				rt2.Weight = &types.Int32Value{Value: 5}

				allRouteTables = v1.RouteTableList{rt1a, rt1b, rt2}
			})

			It("works as expected", func() {

				converted, err := visitor.ConvertVirtualService(vs)

				Expect(err).NotTo(HaveOccurred())
				Expect(converted).To(HaveLen(3))

				Expect(converted[0]).To(WithTransform(getFirstPrefixMatcher, Equal("/foo/a/")))
				Expect(converted[1]).To(WithTransform(getFirstPrefixMatcher, Equal("/foo/b/")))
				Expect(converted[2]).To(WithTransform(getFirstPrefixMatcher, Equal("/foo/")))
			})
		})
```

### LINKED ISSUE #3586: Support Kubernetes-like route table delegation label selector expressions
**Is your feature request related to a problem? Please describe.**
Route table delegation only supports exact label matching at the present time. To allow for more flexible route table selection in complex scenarios, Kubernetes-like label selector expressions should be supported. 

**Describe the solution you'd like**
Support Kubernetes-like expressions as a route table selector attribute.

```yaml
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: 'example'
  namespace: 'gloo-system'
spec:
  virtualHost:
    domains:
    - 'example.com'
    routes:
    - matchers:
       - prefix: '/'
      delegateAction:
        selector:
          # 'expressions' and 'labels' cannot coexist within a single selector 
          expressions:
            - key: domain
              operator: In
              values:
                - example.com
            # Do not include route tables exposing the 'shouldnotexist' label
            - key: shouldnotexist
              operator: !
          namespaces:
          - gloo-system
```
**Additional context**
The idea stemed from #3085 to circumvent awkward workarounds related to route table weights and multiple selectors.

### PR DESCRIPTION
**Pull request by oboukili**
# Description

This PR introduces the ability to use [Kubernetes label selector expressions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-references-in-api-objects) within a virtualservice route delegateAction, allowing for more flexible delegation scenarios.

Closes #3085
Closes #3586

Example virtualservice:

```yaml
apiVersion: gateway.solo.io/v1
kind: VirtualService
metadata:
  name: someservice
  namespace: gloo-system
spec:
  virtualHost:
    domains:
      - some.domain
    routes:
      - matchers:
          - prefix: "/"
        delegateAction:
          selector:
            expressions:
              - 
                 key: foo
                 operator: In
                 values:
                   - bar
                   - baz
              - 
                 key: isitpresent
                 operator: Exists
            namespaces:
              - "*"
```

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3586

### CODE REVIEW

File: projects/gateway/api/v1/virtual_service.proto [hunk 1]
```diff
@@ -245,4 +245,24 @@ message RouteTableSelector {
...
+        DoesNotExist = 7;
+      }
+      // The operator can only be In, NotIn, Equals, DoubleEquals, NotEquals, Exists, or DoesNotExist.
+      Operator operator = 2;
+      repeated string values = 3;
+    }
+
+    // Expressions allow for more flexible Route Tables label matching, such as equality-based requirements, set-based requirements, or a combination of both.
+    // https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#equality-based-requirement
+    repeated Expression expressions = 3;
```
Comment (kdorosh) [inline projects/gateway/api/v1/virtual_service.proto:267]: looks like we can only use expressions or labels, not both.

we should update the comment here and above on `labels` to reflect this. too bad we can't put this in a `oneof` 😢

File: projects/gateway/api/v1/virtual_service.proto [hunk 2]
```diff
@@ -245,4 +245,24 @@ message RouteTableSelector {
     // Delegate to Route Tables whose labels match the ones specified here.
     map<string, string> labels = 2;
 
+    message Expression {
+      string key = 1;
+      enum Operator {
+        UNKNOWN = 0;
```
Comment (kdorosh) [inline projects/gateway/api/v1/virtual_service.proto:251]: why even define the UNKNOWN enum?

File: projects/gateway/api/v1/virtual_service.proto (hunk 1 above)
Comment (oboukili) [reply projects/gateway/api/v1/virtual_service.proto:267]: Quite sad Indeed! I will update the comment, the use case is already handled in the code ans returns a specific warning.

File: projects/gateway/api/v1/virtual_service.proto (hunk 2 above)
Comment (oboukili) [reply projects/gateway/api/v1/virtual_service.proto:251]: You're right, my first intent was to allow for the set based non/existence syntax (key: !labelkey), but the Exists/NotExist operator values take care if that use case, I will update the comments and remove that.

File: docs/content/reference/api/github.com/solo-io/gloo/projects/gateway/api/v1/virtual_service.proto.sk.md [hunk 3]
```diff
@@ -241,13 +243,57 @@ Select route tables for delegation by namespace, labels, or both.
...
+
+---
+### Operator
+
+ 
+Route Table Selector expression operator, while the set-based syntax differs from Kubernetes (kubernetes: `key: !mylabel`, gloo: `key: mylabel, operator: "!"` | kubernetes: `key: mylabel`, gloo: `key: mylabel, operator: exists`), the functionality remains the same.
+
+| Name | Description |
+| ----- | ----------- | 
+| `Equals` |  |
```
[the remark this answers is not in the document]
Comment (kdorosh) [reply docs/content/reference/api/github.com/solo-io/gloo/projects/gateway/api/v1/virtual_service.proto.sk.md]: looks like a solo-kit improvement, opened https://github.com/solo-io/solo-kit/issues/379 to track

**Pull request merged**